File: AoC-2015/Day03/Problem2.py
with open("input.txt", "r") as f:
    line = f.readline().strip()

ans = 1
santa = False
seen = {(0,0)}
sx,sy,rx,ry = 0,0,0,0

for c in line:
    x, y = (sx, sy) if santa else (rx, ry)

    if c == '^':
        y += 1
    elif c == '>':
        x += 1
    elif c == 'v':
        y -= 1
    elif c == '<':
        x -= 1

    if (x,y) not in seen:
        ans += 1
        seen.add((x,y))
    
    if santa:
        sx, sy = x, y
    else:
        rx, ry = x, y

    santa = not santa
    
print(ans)

File: AoC-2025/Day04/Problem2.java
import java.io.FileReader;
import java.io.IOException;
import java.io.BufferedReader;
import java.util.ArrayList;

public class Problem2
{
    public static void main(String[] args) {solve();}

    public static void solve()
    {
        try(BufferedReader br = new BufferedReader(new FileReader("Day04/input.txt")))
        {   
            String line;
            ArrayList<String> grid = new ArrayList<>();
            long ans = 0;

            while((line = br.readLine()) != null) grid.add(line);

            int M = grid.size(), N = grid.get(0).length();
            int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}, {1,1}, {1,-1}, {-1,1}, {-1,-1}};
            ArrayList<ArrayList<Boolean>> seen = new ArrayList<>();

            for(int i=0;i<M;i++){
                ArrayList<Boolean> row = new ArrayList<>();
                for(int j=0;j<N;j++){
                    row.add(false);
                }
                seen.add(row);
            }

            while(true){
                int count = 0;
                for(int i=0;i<M;i++){
                    for(int j=0;j<N;j++){
                        if(grid.get(i).charAt(j) != '@' || seen.get(i).get(j)) continue;

                        int neigh = 0;

                        for(int[] d: dirs){
                            int di = d[0], dj = d[1];

                            if(in(i+di, j+dj, M, N) && grid.get(i+di).charAt(j+dj) == '@' && !seen.get(i+di).get(j+dj)){
                                neigh++;
                            }
                        }

                        if(neigh < 4){
                            count++;
                            seen.get(i).set(j, true);
                        }
                    }
                }
                
                if(count == 0) break;
                ans += count;
            }
                

            System.out.println(ans);
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }

    public static boolean in(int x, int y, int m, int n){ return x >= 0 && y >=0 && x < m && y < n;}
}